sway-ir/src/optimize/dominator.rs

```diff
@@ -0,0 +1,209 @@
+    fn post_order(
+        context: &Context,
+        n: Block,
+        res: &mut PostOrder,
+        on_stack: &mut FxHashSet<Block>,
+        counter: &mut usize,
+    ) {
+        if on_stack.contains(&n) {
+            return;
+        }
+        on_stack.insert(n);
+        for BranchToWithArgs { block: n_succ, .. } in n.successors(context) {
+            post_order(context, n_succ, res, on_stack, counter);
+        }
+        res.block_to_po.insert(n, *counter);
+        res.po_to_block.push(n);
+        *counter += 1;
+    }
+    post_order(context, entry, &mut res, &mut on_stack, &mut counter);
+
+    // We could assert the whole thing, but it'd be expensive.
+    assert!(res.po_to_block.last().unwrap() == &entry);
+
+    res
+}
+
+/// Compute the dominator tree for the CFG.
+pub fn compute_dom_tree(context: &Context, function: &Function) -> (DomTree, PostOrder) {
+    let po = compute_post_order(context, function);
+    let mut dom_tree = DomTree::default();
+    let entry = function.get_entry_block(context);
+
+    // This is to make the algorithm happy. It'll be changed to None later.
+    dom_tree.insert(entry, DomTreeNode::new(Some(entry)));
+    // initialize the dominators tree. This allows us to do dom_tree[b] fearlessly.
+    // Note that we just previously initialized "entry", so we skip that here.
+    for b in po.po_to_block.iter().take(po.po_to_block.len() - 1) {
+        dom_tree.insert(*b, DomTreeNode::new(None));
+    }
+    let mut changed = true;
+
+    while changed {
+        changed = false;
+        // For all nodes, b, in reverse postorder (except start node)
+        for b in po.po_to_block.iter().rev().skip(1) {
+            // new_idom <- first (processed) predecessor of b (pick one)
+            let mut new_idom = b
+                .pred_iter(context)
+                .find(|p| {
+                    // "p" may not be reachable, and hence not in dom_tree.
+                    po.block_to_po
+                        .get(p)
+                        .map_or(false, |p_po| *p_po > po.block_to_po[b])
+                })
+                .cloned()
+                .unwrap();
+            let picked_pred = new_idom;
+            // for all other (reachable) predecessors, p, of b:
+            for p in b
+                .pred_iter(context)
+                .filter(|p| **p != picked_pred && po.block_to_po.contains_key(p))
+            {
+                if matches!(dom_tree[p].parent, Some(_)) {
+                    // if doms[p] already calculated
+                    new_idom = intersect(&po, &mut dom_tree, *p, new_idom);
+                }
+            }
+            let b_node = dom_tree.get_mut(b).unwrap();
+            match b_node.parent {
+                Some(idom) if idom == new_idom => {}
+                _ => {
+                    b_node.parent = Some(new_idom);
+                    changed = true;
+                }
+            }
+        }
+    }
+
+    fn intersect(
```

Comment:
Maybe a comment explaining what `intersect` does.

Reply:
> Maybe a comment explaining what `intersect` does.

It's a bit hard to describe what it does without the context of the paper. Let me give it a try !